File: Objective-C/FlickrSearch/FSResultCell.m
//
//  FSResultCell.m
//  t5
//
//  Created by Alfred Yu on 2020/10/19.
//

#import "FSResultCell.h"
#import "FSPhoto.h"
#import "DataManager.h"

typedef enum enFavoriteStatus
{
    Selected = 0,
    UnSelected
} FavoriteStatus;

@interface FSResultCell () <FlickrPhoto_Delegate>

@property (strong, nonatomic) FSPhoto *photoInfo;
@property (weak, nonatomic) IBOutlet UIImageView *favorite;

@end

@implementation FSResultCell

- (void)awakeFromNib
{
    [super awakeFromNib];
    
    [self setLoadingComplete:NO];
    [self imageViewInit];
    [self favoriteIconInit];
}

- (void) imageViewInit
{
    [self.imageView setClipsToBounds:YES];
    [self.imageView.layer setCornerRadius:9];
    [self.imageView setContentMode:UIViewContentModeScaleAspectFill];
}

- (void) favoriteIconInit
{
    UITapGestureRecognizer *tap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(favIconOnClick)];
    [self.favorite setUserInteractionEnabled:YES];
    [self.favorite addGestureRecognizer:tap];
}

- (void) setLoadingComplete: (BOOL)status
{
    [self.imageView setHidden:(status ? NO : YES)];
    [self.favorite setHidden:(status ? NO : YES)];
    [self.activityView setHidden:(status ? YES : NO)];
    (status ? [self.activityView stopAnimating] : [self.activityView startAnimating]);
    
    if ( [DataManager queryRecordByphotoId:self.photoInfo.photo_id owner:self.photoInfo.photo_owner] )
        [self setCellIsSelectedAsFavorite:YES];
    else
        [self setCellIsSelectedAsFavorite:NO];
}

- (void) loadImageWithInfo: (FSPhoto *)info
{
    self.photoInfo = info;
    
    if ( info.image == nil )
    {
        [self setLoadingComplete:NO];
        
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^
        {
            [self.photoInfo loadImage];
            self.photoInfo.delegate = self;
        });
    }
    else
    {
        [self setPhoto];
    }
}

- (void) setPhoto
{
    [self.imageView setImage:self.photoInfo.image];
    [self setLoadingComplete:YES];
}

- (void) setCellIsSelectedAsFavorite: (BOOL)isSelected
{
    if ( isSelected )
    {
        [self.favorite setImage:[UIImage systemImageNamed:@"heart.fill"]];
        [self.favorite setTag:Selected];
    }
    else
    {
        [self.favorite setImage:[UIImage systemImageNamed:@"heart"]];
        [self.favorite setTag:UnSelected];
    }
}

- (void) switchCellSelected
{
    if ( self.favorite.tag == Selected )
    {
        [DataManager deleteRecordByphotoId:self.label_id.text owner:self.label_owner.text];
        [self setCellIsSelectedAsFavorite:NO];
    }
    else
    {
        [DataManager addRecordByphotoId:self.label_id.text owner:self.label_owner.text source:self.photoInfo.photo_source];
        [self setCellIsSelectedAsFavorite:YES];
    }
}

#pragma mark - FlickrPhoto_Delegate

- (void) photoLoadComplete
{
    dispatch_async(dispatch_get_main_queue(), ^
    {
        [self setPhoto];
    });
}

#pragma mark - gesture event

- (void) favIconOnClick
{
    [self switchCellSelected];
}

@end

File: Swift/FlickrSearch/FSResultCell.swift
//
//  FSResultCell.swift
//  FlickrSearch
//
//  Created by Alfred Yu on 2020/10/21.
//

import UIKit

class FSResultCell: UICollectionViewCell, FlickrPhoto_Delegate
{
    @IBOutlet weak var label_id: UILabel!
    @IBOutlet weak var label_owner: UILabel!
    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var favorite: UIImageView!
    @IBOutlet weak var activityView: UIActivityIndicatorView!
    var photoInfo: FlickrPhoto?
    
    override func awakeFromNib()
    {
        super.awakeFromNib()
        
        setLoadingComplete(status: false)
        imageViewInit()
        
        let tap = UITapGestureRecognizer(target: self, action: #selector(favIconOnClick))
        favorite.isUserInteractionEnabled = true
        favorite.addGestureRecognizer(tap)
        
        favorite.image = UIImage(systemName: "heart")
        favorite.tag = 0
    }

    public func loadImage(photoinfo: FlickrPhoto)
    {
        self.photoInfo = photoinfo;
        
        if photoinfo.image == nil
        {
            setLoadingComplete(status: false)
            
            DispatchQueue.global(qos: .userInitiated).async
            {
                self.photoInfo?.loadImage()
                self.photoInfo?.delegate = self
            }
        }
        else
        {
            setPhoto()
        }
    }
    
    private func imageViewInit()
    {
        imageView.clipsToBounds = true
        imageView.layer.cornerRadius = 9
        imageView.contentMode = .scaleAspectFill
    }
    
    private func setLoadingComplete(status: Bool)
    {
        if status
        {
            imageView.isHidden = false
            favorite.isHidden = false
            activityView.isHidden = true
            activityView.stopAnimating()
        }
        else
        {
            imageView.isHidden = true
            favorite.isHidden = true
            activityView.isHidden = false
            activityView.startAnimating()
        }
    }
    
    private func setPhoto()
    {
        imageView.image = photoInfo?.image
        setLoadingComplete(status: true)
    }
    
    public func photoLoadComplete()
    {
        DispatchQueue.main.async
        {
            self.setPhoto()
        }
    }
    
    @objc private func favIconOnClick()
    {
        if favorite.tag == 0
        {
            favorite.image = UIImage(systemName: "heart.fill")
            favorite.tag = 1
        }
        else
        {
            favorite.image = UIImage(systemName: "heart")
            favorite.tag = 0
        }
    }
}
